File: verify_performance.py
import numpy as np
from scipy.optimize import curve_fit
from pycse import nlinfit
from scipy.stats import pearsonr, spearmanr, kendalltau
import statsmodels.api as sm
from scipy.stats import t

def predict_func(x, params):
    return logistic(x, *params)

def logistic(X,bayta1,bayta2,bayta3,bayta4,bayta5):

    logisticPart = 0.5 - 1 / (1 + np.exp(bayta2 * (X - bayta3)))

    yhat = bayta1 * logisticPart + bayta4 * X + bayta5

    return yhat

def nlpredci(func, xdata, params, residuals, jac):
    alpha = 0.05  # 置信水平
    dof = len(xdata) - len(params)  # 自由度
    tval = t.ppf(1 - alpha / 2, dof)  # t分布值
    popt = params  # 参数估计值

    ypred = func(xdata, *params)  # 预测值
    se = np.sqrt(np.sum(residuals ** 2) / dof)  # 标准误差

    return ypred


def verify_performance(mos, predict_mos):
    predict_mos = predict_mos.flatten()
    mos = mos.flatten()
    print('predict_mos.shape')
    print(predict_mos.shape)
    print('mos.shape')
    print(mos.shape)

    beta = [10, 0, np.mean(predict_mos), 0.1, 0.1]

    # Fitting a curve using the data
    bayta,pcov = curve_fit(logistic, predict_mos, mos, p0 = beta)
    residuals = mos - logistic(predict_mos, *bayta)

    # 计算雅可比矩阵
    jac = np.zeros((len(predict_mos), len(bayta)))

    ypre = logistic(predict_mos, *bayta)


    rmse = np.sqrt(np.sum((ypre - mos) ** 2) / len(mos))  # Root Mean Squared Error
    plcc, _ = pearsonr(mos, ypre)  # Pearson Linear Correlation Coefficient
    srocc, _ = spearmanr(mos, predict_mos)  # Spearman Rank Correlation Coefficient
    krocc, _ = kendalltau(mos, predict_mos)  # Kendall Rank Correlation Coefficient
    #
    return srocc, krocc, plcc, rmse

File: verify_performance.m
%this script is used to calculate the pearson linear correlation
%coefficient and root mean sqaured error after regression

%get the objective scores computed by the IQA metric and the subjective
%scores provided by the dataset

function [srocc,krocc,plcc,rmse] = verify_performance(mos,predict_mos)

predict_mos = predict_mos(:);
mos = mos(:);

%initialize the parameters used by the nonlinear fitting function
beta(1) = 10;
beta(2) = 0;
beta(3) = mean(predict_mos);
beta(4) = 0.1;
beta(5) = 0.1;

%fitting a curve using the data
[bayta,ehat,J] = nlinfit(predict_mos,mos,@logistic,beta);
%given a ssim value, predict the correspoing mos (ypre) using the fitted curve
[ypre,junk] = nlpredci(@logistic,predict_mos,bayta,ehat,J);
% ypre = predict(logistic,fsimValues,bayta,ehat,J);
rmse = sqrt(sum((ypre - mos).^2) / length(mos));%root meas squared error
plcc = corr(mos, ypre, 'type','Pearson'); %pearson linear coefficient
srocc = corr(mos, predict_mos, 'type','spearman');
krocc = corr(mos, predict_mos, 'type','Kendall');

% t = min(predict_mos):0.01:max(predict_mos);
% [ypre junk] = nlpredci(@logistic,t,bayta,ehat,J);
% p = plot(predict_mos,mos,'+');
% set(p,'Color','blue','LineWidth',1);
% hold on;
% p = plot(t,ypre);
% set(p,'Color','black','LineWidth',2);
% hhh=legend('Images in NNID','Curve fitted with logistic function', 'Location','NorthEast','FontSize',16);
% set(hhh,'Fontsize',14);
% xlabel('Objective score by Proposed','FontSize',16);
% ylabel('DMOS','FontSize',16);
% print(1, '-dpng', 'ssim');
end
